runtime/sema/check_transaction_declaration.go

```diff
@@ -74,6 +74,28 @@ func (checker *Checker) VisitTransactionDeclaration(declaration *ast.Transaction
 		checker.checkRolePrepareParameterList(roleDeclaration, transactionType)
 	}
 
+	transactionPrepareParameterCount := len(transactionType.PrepareParameters)
+
+	// If the transaction has roles and the prepare block has parameters,
+	// then there should be exactly one role for each parameter
+	roleCount := len(declaration.Roles)
+	if roleCount > 0 &&
+		prepareFunctionDeclaration != nil &&
+		transactionPrepareParameterCount > 0 &&
+		roleCount != transactionPrepareParameterCount {
```

Comment:
Do they have to be equal? `roleCount <= transactionPrepareParameterCount` should be also fine?
i.e: There can be accounts used by the transaction's prepare only, but not by any roles.

But maybe this should be also something to discuss on the FLIP